engine/collection/synchronization/engine.go

```diff
@@ -54,310 +67,259 @@ func New(
+		msg, ok := e.pendingSyncResponses.Get()
+		if ok {
+			e.onSyncResponse(msg.OriginID, msg.Payload.(*messages.SyncResponse))
+			e.metrics.MessageHandled(metrics.EngineClusterSynchronization, metrics.MessageSyncResponse)
 			continue
 		}
-		if err != nil {
-			return fmt.Errorf("could not get block by ID (%s): %w", blockID, err)
+
+		msg, ok = e.pendingBlockResponses.Get()
+		if ok {
+			e.onBlockResponse(msg.OriginID, msg.Payload.(*messages.ClusterBlockResponse))
+			e.metrics.MessageHandled(metrics.EngineClusterSynchronization, metrics.MessageBlockResponse)
+			continue
 		}
-		blocks = append(blocks, block)
-	}
 
-	// if there are no blocks to send, skip network message
-	if len(blocks) == 0 {
-		e.log.Debug().Msg("skipping empty batch response")
-		return nil
+		// when there is no more messages in the queue, back to the loop to wait
+		// for the next incoming message to arrive.
+		return
 	}
+}
 
-	// send the response
-	res := &messages.ClusterBlockResponse{
-		Nonce:  req.Nonce,
-		Blocks: blocks,
-	}
-	err := e.conduit.Unicast(res, originID)
+// onSyncResponse processes a synchronization response.
+func (e *Engine) onSyncResponse(originID flow.Identifier, res *messages.SyncResponse) {
+	final, err := e.state.Final().Head()
 	if err != nil {
-		return fmt.Errorf("could not send batch response: %w", err)
+		e.log.Error().Err(err).Msg("could not get last finalized header")
+		return
 	}
-
-	e.metrics.MessageSent(metrics.EngineClusterSynchronization, metrics.MessageBlockResponse)
-
-	return nil
+	e.core.HandleHeight(final, res.Height)
 }
 
 // onBlockResponse processes a response containing a specifically requested block.
-func (e *Engine) onBlockResponse(originID flow.Identifier, res *messages.ClusterBlockResponse) error {
-
+func (e *Engine) onBlockResponse(originID flow.Identifier, res *messages.ClusterBlockResponse) {
 	// process the blocks one by one
 	for _, block := range res.Blocks {
-		e.processIncomingBlock(originID, block)
-	}
-
-	return nil
-}
-
-// processIncoming processes an incoming block, so we can take into account the
-// overlap between block IDs and heights.
-func (e *Engine) processIncomingBlock(originID flow.Identifier, block *clustermodel.Block) {
-
-	shouldProcess := e.core.HandleBlock(block.Header)
-	if !shouldProcess {
-		return
-	}
-
-	synced := &events.SyncedClusterBlock{
-		OriginID: originID,
-		Block:    block,
+		if !e.core.HandleBlock(block.Header) {
+			continue
+		}
+		synced := &events.SyncedClusterBlock{
+			OriginID: originID,
+			Block:    block,
+		}
+		e.comp.SubmitLocal(synced)
 	}
-
-	e.comp.SubmitLocal(synced)
 }
 
 // checkLoop will regularly scan for items that need requesting.
 func (e *Engine) checkLoop() {
-	poll := time.NewTicker(e.pollInterval)
+	pollChan := make(<-chan time.Time)
+	if e.pollInterval > 0 {
+		poll := time.NewTicker(e.pollInterval)
+		pollChan = poll.C
+		defer poll.Stop()
+	}
 	scan := time.NewTicker(e.scanInterval)
```

Comment:
Any reason not to treat `pollInterval` and `scanInterval` the same here? IE handle negative values, and `defer <ticker>.Stop()` for both?

Reply:
For execution nodes consensus sync engine passed 0, meaning it doesn't want to poll height. 
I don't know if it's needed for collection nodes but no harm in keeping it I guess

Reply:
should we add the `defer scanStop()` though?